auditing/timescaledb.go

```diff
@@ -0,0 +1,360 @@
+		maxIdleConns = *tc.MaxIdleConns
+	}
+	a.db.SetMaxIdleConns(maxIdleConns)
+
+	connMaxLifetime := 2 * time.Minute
+	if tc.ConnMaxLifetime != nil {
+		connMaxLifetime = *tc.ConnMaxLifetime
+	}
+	a.db.SetConnMaxLifetime(connMaxLifetime)
+
+	maxOpenConns := 95
+	if tc.MaxOpenConns != nil {
+		maxOpenConns = *tc.MaxOpenConns
+	}
+	a.db.SetMaxOpenConns(maxOpenConns)
+
+	err = a.initialize()
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize timescaledb backend: %w", err)
+	}
+
+	a.log.Info("connected to timescaledb backend")
+
+	return a, nil
+}
+
+func (a *timescaleAuditing) initialize() error {
+	type txStatement struct {
+		query string
+		args  []any
+	}
+
+	initialSchema := &migrator.Migration{
+		Name: "Initial database schema",
+		Func: func(tx *sql.Tx) error {
+			for _, stmt := range []txStatement{
+				{
+					query: `CREATE EXTENSION IF NOT EXISTS timescaledb`,
+				},
+				{
+					query: `CREATE EXTENSION IF NOT EXISTS pg_stat_statements`,
+				},
+				{
+					query: `CREATE TABLE IF NOT EXISTS traces (
+						timestamp timestamp NOT NULL,
+						entry jsonb NOT NULL,
+						ts tsvector GENERATED ALWAYS AS (to_tsvector('simple', entry)) STORED
+					)`,
+				},
+				{
+					query: `SELECT create_hypertable('traces', 'timestamp', chunk_time_interval => $1::interval, if_not_exists => TRUE)`,
+					args:  []any{a.config.ChunkInterval},
+				},
+				{
+					query: `ALTER TABLE traces SET (
+						timescaledb.compress,
+						timescaledb.compress_orderby = 'timestamp'
+					)`,
+				},
+				{
+					query: `SELECT add_compression_policy('traces', $1::interval)`,
+					args:  []any{a.config.CompressionInterval},
+				},
+				{
+					query: `CREATE INDEX IF NOT EXISTS traces_gin_idx ON traces USING GIN (entry)`,
+				},
+				{
+					query: `CREATE INDEX IF NOT EXISTS ts_idx ON traces USING GIN (ts)`,
+				},
+				{
+					query: `SELECT add_retention_policy('traces', $1::interval)`,
+					args:  []any{a.config.Retention},
+				},
+			} {
+				if _, err := tx.Exec(stmt.query, stmt.args...); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+	}
+
+	m, err := migrator.New(
+		migrator.WithLogger(migrator.LoggerFunc(func(msg string, args ...interface{}) {
+			a.log.Info(fmt.Sprintf(msg, args...))
+		})),
+		migrator.Migrations(
+			initialSchema,
+		),
+	)
+	if err != nil {
+		return err
+	}
+
+	if err := m.Migrate(a.db.DB); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a *timescaleAuditing) Flush() error {
+	return nil
+}
+
+func (a *timescaleAuditing) Index(entry Entry) error {
+	if entry.Timestamp.IsZero() {
+		return errors.New("timestamp is not set")
+	}
+
+	q := "INSERT INTO traces (timestamp, entry) VALUES (:timestamp, :entry)"
+
+	e, err := json.Marshal(entry)
+	if err != nil {
+		return fmt.Errorf("error marshaling entry: %w", err)
+	}
+
+	row := timescaledbRow{
+		Timestamp: entry.Timestamp.UTC(),
+		Entry:     e,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
```

Comment:
can you create a const at the top of the file for `2*time.Second`?

Reply:
Done!